src/main/java/appeng/tile/powersink/AEBasePoweredTile.java

```diff
@@ -19,7 +19,308 @@
+	{
+		return PowerUnits.AE.convertTo( externalUnit, Math.max( 0.0, this.getFunnelPowerDemand( externalUnit.convertTo( PowerUnits.AE, maxPowerRequired ) ) ) );
+	}
+
+	protected double getFunnelPowerDemand( final double maxRequired )
+	{
+		return this.getInternalMaxPower() - this.getInternalCurrentPower();
+	}
+
+	@Override
+	public final double injectExternalPower( final PowerUnits input, final double amt )
+	{
+		return PowerUnits.AE.convertTo( input, this.funnelPowerIntoStorage( input.convertTo( PowerUnits.AE, amt ), Actionable.MODULATE ) );
+	}
+
+	protected double funnelPowerIntoStorage( final double power, final Actionable mode )
+	{
+		return this.injectAEPower( power, mode );
+	}
+
+	@Override
+	public final double injectAEPower( double amt, final Actionable mode )
+	{
+		if( amt < 0.000001 )
+		{
+			return 0;
+		}
+
+		if( mode == Actionable.SIMULATE )
+		{
+			final double fakeBattery = this.getInternalCurrentPower() + amt;
+
+			if( fakeBattery > this.getInternalMaxPower() )
+			{
+				return fakeBattery - this.getInternalMaxPower();
+			}
+
+			return 0;
+		}
+		else
+		{
+			if( this.getInternalCurrentPower() < 0.01 && amt > 0.01 )
+			{
+				this.PowerEvent( PowerEventType.PROVIDE_POWER );
+			}
+
+			this.setInternalCurrentPower( this.getInternalCurrentPower() + amt );
+			if( this.getInternalCurrentPower() > this.getInternalMaxPower() )
+			{
+				amt = this.getInternalCurrentPower() - this.getInternalMaxPower();
+				this.setInternalCurrentPower( this.getInternalMaxPower() );
+				return amt;
+			}
+
+			return 0;
+		}
+	}
+
+	protected void PowerEvent( final PowerEventType x )
+	{
+		// nothing.
+	}
+
+	@Override
+	public final double getAEMaxPower()
+	{
+		return this.getInternalMaxPower();
+	}
+
+	@Override
+	public final double getAECurrentPower()
+	{
+		return this.getInternalCurrentPower();
+	}
+
+	@Override
+	public final boolean isAEPublicPowerStorage()
+	{
+		return this.isInternalPublicPowerStorage();
+	}
+
+	@Override
+	public final AccessRestriction getPowerFlow()
+	{
+		return this.getInternalPowerFlow();
+	}
+
+	@Override
+	public final double extractAEPower( final double amt, final Actionable mode, final PowerMultiplier multiplier )
+	{
+		return multiplier.divide( this.extractAEPower( multiplier.multiply( amt ), mode ) );
+	}
+
+	protected double extractAEPower( double amt, final Actionable mode )
+	{
+		if( mode == Actionable.SIMULATE )
+		{
+			if( this.getInternalCurrentPower() > amt )
+			{
+				return amt;
+			}
+			return this.getInternalCurrentPower();
+		}
+
+		final boolean wasFull = this.getInternalCurrentPower() >= this.getInternalMaxPower() - 0.001;
+		if( wasFull && amt > 0.001 )
+		{
+			this.PowerEvent( PowerEventType.REQUEST_POWER );
+		}
+
+		if( this.getInternalCurrentPower() > amt )
+		{
+			this.setInternalCurrentPower( this.getInternalCurrentPower() - amt );
+			return amt;
+		}
+
+		amt = this.getInternalCurrentPower();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Introduce a new variable instead of reusing the parameter "amt". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1226)